### ISSUE TITLE
List of future optimization

### ISSUE THREAD

**Issue by xmyqsh** [traveller59/spconv#154]
1. voxelization generation
2. max/avg pool
3. normal sparse conv unrolling
4. gather/scatter time benchmark, fused gemm by hand(Tiled Matrix Multiplication by CUDA) or fused gemm by CUTLASS or any other opt by CUTLASS

1 is the main bottleneck currently. 1 and 2 could be optimized together.
3 is the bottleneck in conv, aka, stride=2 normal sparse conv.
4 is not bottleneck in spconv. Further optimization need benchmark further.

According to Cannikin Law, we should do 1,2 first, which is easy to implement by Atomic operation.

**Comment by traveller59** [comment]
Why the gather/scatter isn't bottleneck? DOPS only takes 12ms on total UNet but spconv takes 40ms on encoder.

**Comment by xmyqsh** [comment]
DOPS could be faster when it uses spconv :)
https://github.com/traveller59/spconv/issues/155
Let's waiting for its open source.

**Comment by aaroey** [comment]
DOPS is not using spconv at the moment. We did some simple benchmark in the past, and found that spconv is faster with larger inputs/filters and our implementation is better with smaller inputs/filters. So we're planning to switch to spconv.

**Comment by traveller59** [comment]
@aaroey according to DOPS paper, your implementation takes only 12ms in a complete U-Net but spconv takes [40ms](https://github.com/traveller59/spconv/blob/develop/test/benchmark.py) in encoder of U-Net with waymo data. is there anything wrong with this conclusion?

**Comment by afathi3** [comment]
@traveller59 I am not sure if we are comparing apples to apples. I think number of layers in encoder and decoder, number of conv channels, number of points, etc. affect the speed obviously. Also the GPU matters of course (we benchmarked on a Titan V).

**Comment by xmyqsh** [comment]
Thanks all.
Hoping for someone merges gather/scatter into gemm efficiently.

**Comment by xmyqsh** [comment]
@traveller59 @aaroey @afathi3 
Look at MinkowskiEngine's [convolution.cu](https://github.com/StanfordVL/MinkowskiEngine/blob/master/src/convolution.cu)!
A non-typical tiled matrix multiplication which is index of index addressing friendly.

**Comment by traveller59** [comment]
@xmyqsh 
it looks same as SparseConvNet...

**Comment by xmyqsh** [comment]
@traveller59 No!
I will test it.

**Comment by xmyqsh** [comment]
@traveller59 
They are almost the same.
But two differences make them separate.
First, for n_Hot, loop in kernel or loop outside kernel, the latter is better.
Second, for dim, `dim(n_active, out_channel)` or `dim(out_channel, n_active)`, the latter seems better.

Also, I have added a `NumILP` support for [MinkowskiEngine](https://github.com/StanfordVL/MinkowskiEngine/pull/164), it looks much clear and efficient on the accessing to the rulebook than SparseConvNet, because of the choice of `dim(out_channel, n_active)` and loop outside kernel.

For the accessing of the index of index `inFeatures` aka, `I[ty[v]][tx] = inFeatures[R0[v] * input_stride + tx];`, I don't think it disturbs the SIMT now, except for some corner case, which could be handled by further tuning the code.

**Comment by traveller59** [comment]
so MinkowskiEngine is very slow when number of channels is large too?
I add the SparseConvNet fused conv to this project, you can select it by passing ```algo=spconv.ConvAlgo.SparseConvNet``` to bench it or bench MinkowskiEngine in spconv codebase.

**Comment by xmyqsh** [comment]
Perfect! So nice!
According to my coarse estimation in the wild, `MinkowskiEngine` could achieve 100ms in your benchmark with `bn` with `relu` without `NumILP` opt without `corner case` opt. It is much faster than `SparseConvNet`.

I will release further benchmark results here.

**Comment by traveller59** [comment]
I add the [Minkowski conv](https://github.com/traveller59/spconv/blob/develop/src/spconv/fused_conv.cu#L58) kernel to spconv. it's really fast.

**Comment by xmyqsh** [comment]
@traveller59 
Wonderful! 50ms without `bn` without `relu`!
I'm working on `NumILP` opt and `corner case` opt.

**Comment by xmyqsh** [comment]
Test report:

Minkowski: 50ms

For `NumILP`:
with NumILP = shared_mem_size/4: 168ms
with NumILP = shared_mem_size/8: 97ms

For `corner case`:
with output_ch = output_ch + 1: 78ms

Conclusion:

`NumILP` is used to hide memory access latency and computation latency. It is worked in gather/scatter which is memory copy intensive kernel, but not worked with `matmul` which implies that the computation latency is not very large with the latest CUDA toolkit, and the memory copy latency is relieved largely by the shared memory.

`corner case` is not optimized by `Minkowski` which has irregular output_channel. But it doesn't matter, because the typical network's channel is divisible by 8. And, it is the reason to make `out_ channel` in block.x .

**Comment by traveller59** [comment]
Thanks for your test.
I am still working on gemm and cutlass. the Minkowski fused conv only support float/double,  we need to develop a unified software to handle SIMT, tensor op and other new gpu features like cutlass does.

**Comment by xmyqsh** [comment]
@traveller59 I have reproduced DOPS's 12ms! @aaroey is correct. spconv is larger inputs/filters/channels friendly.

I also tested [my new tiled matrix multiplication algorithm](https://github.com/StanfordVL/MinkowskiEngine/issues/166), it slightly slower than MinkowskiEngine. But it has the property of 
```
access in Indice M*N times
access out indice M*N times
use atomicAdd  M*N*K/shared_memory_size times
```
which makes it possible to merge gather/gemm/scatter into one kernel by cutlass or a customed cutlass which will help spconv small inputs/filters/channels friendly.
